src/main.cpp

```diff
@@ -20,35 +18,55 @@ static std::string joined_board;
 static std::string sequence_str;
 static std::vector<std::vector<std::string>> levels;
 
-extern "C" {
-
-void sokoban_initialize() {
-    /* Read files from directory */
-    const std::string source_path = "src/assets/levels";
-    std::filesystem::path levels_dir = std::filesystem::directory_entry(source_path);
+void read_levels(const std::string path = "src/assets/levels") {
+    std::map<int, std::vector<std::string>> ordered_levels;
+    std::filesystem::path levels_dir =
+        std::filesystem::directory_entry(path);
 
     for (const auto& entry : std::filesystem::directory_iterator(levels_dir)) {
         std::ifstream level_file(entry.path());
-        std::string line;
-        std::vector<std::string> level;
 
         if (!level_file) {
             throw std::invalid_argument("Cannot open file");
         }
 
-        std::regex valid_elems("[#@$*.+]+");
-        while (std::getline(level_file, line)) {
-            if (std::regex_search(line, valid_elems)) {
+        std::regex soko_elems_reg("[#@$*.+ ]+");
+        std::vector<std::string> level;
+
+        for (std::string line; std::getline(level_file, line);) {
+            if (std::regex_match(line, soko_elems_reg)) {
                 level.push_back(line);
             }
             else if (!level.empty()) {
                 break;
             }
         }
+
+        if (level.size() > 2) {
+            std::string path = entry.path().u8string();
+            std::regex level_num_reg("(\\d+)\\.xsb$");
+            std::smatch match;
+
+            if (std::regex_search(path, match, level_num_reg)) {
+                ordered_levels[std::stoi(match[1])] = level;
```

Comment:
Should we use match.at(1)?

Reply:
Probably not necessary, since if the match is truthy it has to fill in the capturing group. But on the other hand, can't hurt either, so feel free to commit it directly to main.